### DISCUSSION TITLE
Improving performance on Apple Silicon chips

### DISCUSSION THREAD

**Discussion by petersterne** [Stage-Whisper/Stage-Whisper#31] [Ideas]
We've discussed this a bit on the Discord, but there's currently a bug in PyTorch that prevents whisper from using the M1 GPU.

Here's the bug: https://github.com/pytorch/pytorch/issues/82645

Until that bug is fixed, whisper can only run on the cpu. That makes it prohibitively difficult to run the larger whisper models, which may be required for accurate transcription of non-English languages. (English transcription is usually pretty accurate even with the smaller models, which run acceptably on the CPU.)

One possible fix is CPU quantization: https://spell.ml/blog/pytorch-quantization-X8e7wBAAACIAHPhT

On Discord, [rags from pyq](https://github.com/amanpyq) said he is currently looking into the possibility of quantizing whisper's ML models to increase CPU performance.

**Comment by BurntNorton42** [comment]
thanks for this post. I thought I had done something wrong as my M1 Max MacBook with 64GB ram and 2TB SSD was taking an hour to transcribe 2 mins of recording! I'll keep an eye on the GitHub for developments.

**Comment by harrislapiroff** [comment]
From a UI perspective we could do some expectations management in the interface and bias people toward the smaller models. Some ideas:

- Instead of just labeling models tiny, base, small, medium, large we could use labels like Tiny (fast), Base (standard), Small (accurate), Medium (more accurate), Large (most accurate)
- We could try to add order of magnitude time estimates, maybe based on our own testing on different architectures

**Reply by harrislapiroff** [reply]
Though I guess these labels might only be correct for English since it seems like the model is less accurate for other languages generally

**Reply by oenu** [reply]
<img width="322" alt="image" src="https://user-images.githubusercontent.com/51684443/193090487-28c1a54b-cc02-44ca-bf20-b63289245ace.png">
I have implemented a description of each model, I can change the colour/iconography around to give green/blue for tiny/base models,  orange for small/medium models, and red for Large models. 

(the copy in the dialog box is 100% not final, just something to fill the space before deployment)

If you would like to suggest descriptions I can drop them in the next PR

**Reply by harrislapiroff** [reply]
Looks great, @oenu

**Reply by oenu** [reply]
@harrislapiroff Just did some digging and you can actually check free system memory / total system memory / cpu information [from electron](https://ourcodeworld.com/articles/read/108/how-to-get-operative-system-and-hardware-information-with-electron-framework#:~:text=Get%20amount%20of,os.freemem%20method.)

If we know that M1 doesnt perform well I could run a regex pass over the CPU to detect those chips and give warnings when installing / every time a user tries to run a large model

**Reply by harrislapiroff** [reply]
@oenu I think I want more subtlety than a warning. I don't want users to feel bad for picking a larger model—I want them to

1. know it's OK to pick a smaller one (without description, your instinct might be "well of course I want medium" so I like that you said Base is the default model) and
3. not be surprised when picking a larger one takes a long time, which is why providing time estimates would be great, even if they were very broad ranges (30m to an hour)—but I realize even that will be complicated without a lot of testing on our part.

So for now I think your solution looks good and I'll probably do some more thinking about it for another iteration.

**Reply by oenu** [reply]
@harrislapiroff Sounds good, I like the path you are heading down. I think describing use cases might be an avenue to explore. From my limited experiments the tiny/dedicated english models are more than sufficient for interview style content. 

- Tiny/base for American/British RP accents
- base/small for foreign language translation tasks, english song lyrics, some of the more difficult accents
- medium for song lyrics in foreign languages
- large for highly complex audio files

In the meantine I have added a setting to the settings panel that locks a user out from selecting a large model. I will add a little disclaimer that they have to accept before we allow them to give over 10+gigs of sysram to a (for now) orphaned python process

**Reply by petersterne** [reply]
It's not that M1 doesn't perform well compared to other CPUs. It's just that M1 should be able to run whisper on the GPU but currently can't.
